### PR TITLE
end2endtest: refactor and extend

### PR DESCRIPTION
**Pull request by altergui**
I'll be pushing here my WIP on end2endtest package

- apiclient: named returns
- apiclient: add context to all the Wait* methods
- end2endtest: use api.Wait* methods, dropping duplicated code in helpers.go

### CODE REVIEW

File: apiclient/election.go [hunk 1]
```diff
@@ -30,8 +30,9 @@ func (c *HTTPclient) Election(electionID types.HexBytes) (*api.Election, error)
 	return election, nil
 }
 
-// NewElectionrRaw creates a new election given the protobuf Process message.
-func (c *HTTPclient) NewElectionRaw(process *models.Process) (types.HexBytes, error) {
+// NewElectionRaw creates a new election given the protobuf Process message.
+func (c *HTTPclient) NewElectionRaw(
+	process *models.Process) (electionID types.HexBytes, err error) {
```
Comment (p4u) [inline apiclient/election.go]: Don't use named returns, please. Only if its completely necessary.

**Comment by altergui** [comment]
i'm starting to make a structure, inspired in a mixture of `cmd/node` and `cmd/vochaintest`
any positive or negative feedback appreciated, will continue tomorrow

**Comment by p4u** [comment]
Hey @altergui see this new option now available for the Gateway API and Vocone.

https://github.com/vocdoni/vocdoni-node/blob/master/dockerfiles/testsuite/env.gateway0#L6

The gateway vocdoni address must be funded, you can do that in the genesis file

https://github.com/vocdoni/vocdoni-node/blob/master/dockerfiles/testsuite/genesis.json#L191

The URL for querying the faucet would be: `http://gateway:9090/v2/faucet/dev/<address>`

You can use this helper function from the apiclient https://github.com/vocdoni/vocdoni-node/blob/master/apiclient/helpers.go#L142 (just set token as empty string)

Note that the `dev` word on the faucet URL is the gateway's CHAIN flag or env var. 

Same applies for Vocone https://github.com/vocdoni/vocdoni-node/blob/master/dockerfiles/vocone/.env#L21

Enjoy!

**Comment by altergui** [comment]
amazing @p4u 

here's some progress from today
```
./end2endtest --host http://127.0.0.1:9090/v2 --faucet=http://127.0.0.1:9090/v2/faucet/dev/
```

i managed to get to this point

```
2023-01-05T12:40:00-03:00	INFO	end2endtest/main.go:152	getting faucet package
2023-01-05T12:40:00-03:00	INFO	end2endtest/main.go:164	creating Vocdoni account 0xc837D533c286D769c095a60Bbe9baF08F4b25c03
2023-01-05T12:40:00-03:00	DEBUG	end2endtest/main.go:165	faucetPackage is 7061796c6f61643a225c7830385c783831dea05c7866355c7830335c7831325c7831345c786338375c786435335c75303038365c786437695c7863305c7839355c7861365c7830625c7862655c7839625c7861665c7830385c7866345c7862325c5c5c7830335c783138642220207369676e61747572653a22635c786135355c7866615c783937605c7865635c786433365c7861625c7866627b5c786239785c7831663a445c783766395c7830655c7839345c786139765c786264365c783833426573605c7864344c5c7831395c7863365c783166305c7839345c78643542425c7861375c7838394d315c783863525c7830302f5c7866635c7863616c5c7862335c7865665c7866303d5c7865325c786163265c7864315c7830355c7830325c78383267465c78303022
2023-01-05T12:40:00-03:00	DEBUG	data/ipfs.go:347	computed cid: bagaaierabrdqc7gsbtmgeerwtwlsgbzp7xa7a4fs7ue5bjyv6vxhibwqus5q
2023-01-05T12:40:00-03:00	DEBUG	apiclient/client.go:139	POST http://127.0.0.1:9090/v2/accounts
2023-01-05T12:40:00-03:00	DEBUG	apiclient/client.go:139	GET http://127.0.0.1:9090/v2/chain/transactions/reference/aadfcd8f435cc6c0be1e3585cf1eb23949631899c6fa6823b4bed942e0d8ef91
2023-01-05T12:40:04-03:00	DEBUG	apiclient/client.go:139	GET http://127.0.0.1:9090/v2/chain/transactions/reference/aadfcd8f435cc6c0be1e3585cf1eb23949631899c6fa6823b4bed942e0d8ef91
2023-01-05T12:40:08-03:00	DEBUG	apiclient/client.go:139	GET http://127.0.0.1:9090/v2/chain/transactions/reference/aadfcd8f435cc6c0be1e3585cf1eb23949631899c6fa6823b4bed942e0d8ef91
2023-01-05T12:40:12-03:00	DEBUG	apiclient/client.go:139	GET http://127.0.0.1:9090/v2/chain/transactions/reference/aadfcd8f435cc6c0be1e3585cf1eb23949631899c6fa6823b4bed942e0d8ef91
2023-01-05T12:40:16-03:00	DEBUG	apiclient/client.go:139	GET http://127.0.0.1:9090/v2/chain/transactions/reference/aadfcd8f435cc6c0be1e3585cf1eb23949631899c6fa6823b4bed942e0d8ef91
2023-01-05T12:40:20-03:00	DEBUG	apiclient/client.go:139	GET http://127.0.0.1:9090/v2/chain/transactions/reference/aadfcd8f435cc6c0be1e3585cf1eb23949631899c6fa6823b4bed942e0d8ef91
2023-01-05T12:40:24-03:00	DEBUG	apiclient/client.go:139	GET http://127.0.0.1:9090/v2/chain/transactions/reference/aadfcd8f435cc6c0be1e3585cf1eb23949631899c6fa6823b4bed942e0d8ef91
2023-01-05T12:40:28-03:00	DEBUG	apiclient/client.go:139	GET http://127.0.0.1:9090/v2/chain/transactions/reference/aadfcd8f435cc6c0be1e3585cf1eb23949631899c6fa6823b4bed942e0d8ef91
2023-01-05T12:40:32-03:00	DEBUG	apiclient/client.go:139	GET http://127.0.0.1:9090/v2/chain/transactions/reference/aadfcd8f435cc6c0be1e3585cf1eb23949631899c6fa6823b4bed942e0d8ef91
2023-01-05T12:40:36-03:00	DEBUG	apiclient/client.go:139	GET http://127.0.0.1:9090/v2/chain/transactions/reference/aadfcd8f435cc6c0be1e3585cf1eb23949631899c6fa6823b4bed942e0d8ef91
2023-01-05T12:40:40-03:00	FATAL	end2endtest/main.go:177	gave up waiting for tx aadfcd8f435cc6c0be1e3585cf1eb23949631899c6fa6823b4bed942e0d8ef91 to be mined: context deadline exceeded
```

checking the logs i see my testsuite is not producing blocks :facepalm: but i'm done for today so i'll dig into this tomorrow

**Comment by altergui** [comment]
so, it's basically working, but sadly not yet ready for integrating into start_test.sh since it's timing out, and i'm also timed out today :rofl:, i'll let someone else take it from here,
to even begin with, i'd need to dig into why the (new zerolog) log format is not showing timestamps

https://github.com/vocdoni/vocdoni-node/actions/runs/3856145237/jobs/6571998051#step:3:6149

```
### Logs from failed test (end2endtest) ###
3:16PM INF go.vocdoni.io/dvote/cmd/end2endtest/main.go:81 > logger construction succeeded at level info with output stdout
3:16PM INF go.vocdoni.io/dvote/cmd/end2endtest/main.go:87 > new account generated, private key is [7bff39fefc01ca0f34bbcd664708eddcc5fd93587c68d00792849a9ca60e58cb]
3:16PM INF go.vocdoni.io/dvote/cmd/end2endtest/main.go:152 > getting faucet package
3:16PM INF go.vocdoni.io/dvote/cmd/end2endtest/main.go:164 > creating Vocdoni account 0x50597b9F6Dd5e9ee264C1A831AEdf65B3B14d4a0
3:16PM INF go.vocdoni.io/dvote/cmd/end2endtest/main.go:189 > account 0x50597b9F6Dd5e9ee264C1A831AEdf65B3B14d4a0 balance is 100
3:16PM INF go.vocdoni.io/dvote/cmd/end2endtest/main.go:196 > new census created with id f614e8a8facd3f19926c2a488b1ede95a69893cc446783db29741ba063f054d6
3:16PM INF go.vocdoni.io/dvote/cmd/end2endtest/main.go:213 > added 10 participants to census f614e8a8facd3f19926c2a488b1ede95a69893cc446783db29741ba063f054d6
3:16PM INF go.vocdoni.io/dvote/cmd/end2endtest/main.go:227 > census f614e8a8facd3f19926c2a488b1ede95a69893cc446783db29741ba063f054d6 size is 10
3:16PM INF go.vocdoni.io/dvote/cmd/end2endtest/main.go:234 > census published with root 21bd9e0b0b2264e3fe6a352999388a64e6d538e07125a43d58a703b727936d5b
3:16PM INF go.vocdoni.io/dvote/cmd/end2endtest/main.go:266 > generating 2 voting proofs
3:16PM INF go.vocdoni.io/dvote/cmd/end2endtest/main.go:266 > generating 2 voting proofs
3:16PM INF go.vocdoni.io/dvote/cmd/end2endtest/main.go:266 > generating 2 voting proofs
3:16PM INF go.vocdoni.io/dvote/cmd/end2endtest/main.go:266 > generating 2 voting proofs
3:16PM INF go.vocdoni.io/dvote/cmd/end2endtest/main.go:266 > generating 2 voting proofs
3:16PM INF go.vocdoni.io/dvote/cmd/end2endtest/main.go:341 > created new election with id c5d2460186f750597b9f6dd5e9ee264c1a831aedf65b3b14d4a0020000000000 - now wait until it starts
3:16PM INF go.vocdoni.io/dvote/apiclient/helpers.go:126 > waiting for election c5d2460186f750597b9f6dd5e9ee264c1a831aedf65b3b14d4a0020000000000 to reach status READY
3:16PM INF go.vocdoni.io/dvote/cmd/end2endtest/main.go:358 > sending 2 votes
3:16PM INF go.vocdoni.io/dvote/cmd/end2endtest/main.go:358 > sending 2 votes
3:16PM INF go.vocdoni.io/dvote/cmd/end2endtest/main.go:358 > sending 2 votes
3:16PM INF go.vocdoni.io/dvote/cmd/end2endtest/main.go:358 > sending 2 votes
3:16PM INF go.vocdoni.io/dvote/cmd/end2endtest/main.go:358 > sending 2 votes
3:16PM INF go.vocdoni.io/dvote/cmd/end2endtest/main.go:396 > successfully sent 2 votes
3:16PM INF go.vocdoni.io/dvote/cmd/end2endtest/main.go:396 > successfully sent 2 votes
3:16PM INF go.vocdoni.io/dvote/cmd/end2endtest/main.go:396 > successfully sent 2 votes
3:16PM INF go.vocdoni.io/dvote/cmd/end2endtest/main.go:396 > successfully sent 2 votes
3:16PM INF go.vocdoni.io/dvote/cmd/end2endtest/main.go:396 > successfully sent 2 votes
3:16PM INF go.vocdoni.io/dvote/cmd/end2endtest/main.go:411 > 10 votes submitted successfully, took 2.007450036s (4 votes/second)
3:16PM INF go.vocdoni.io/dvote/cmd/end2endtest/main.go:415 > waiting for all the votes to be registered...
3:16PM INF go.vocdoni.io/dvote/cmd/end2endtest/main.go:425 > verified 0/10 votes
3:16PM INF go.vocdoni.io/dvote/cmd/end2endtest/main.go:431 > 10 votes registered successfully, took 7.009779174s (1 votes/second)
3:16PM INF go.vocdoni.io/dvote/cmd/end2endtest/main.go:440 > ending election...
3:16PM INF go.vocdoni.io/dvote/cmd/end2endtest/main.go:460 > election c5d2460186f750597b9f6dd5e9ee264c1a831aedf65b3b14d4a0020000000000 status is ENDED
3:16PM INF go.vocdoni.io/dvote/apiclient/helpers.go:126 > waiting for election c5d2460186f750597b9f6dd5e9ee264c1a831aedf65b3b14d4a0020000000000 to reach status RESULTS
Error: Process completed with exit code 127.
```

**Comment by altergui** [comment]
i'd vote to merge the progress so far, without the last commit 2632eb1935ef4c9ea2574537688c48446c9f6542 that integrates end2endtest into start_test.sh and would break CI, because all the previous commits already add value, can be worked on top of this, and if we don't merge it now, it will mean extra work in the future for rebasing

i'm rewinding the branch 1 commit (HEAD~) to 8833ba92dd0cc240f65a1517003ce14f1b614d89

**Comment by altergui** [comment]
i.e. ready to review and merge

**Comment by p4u** [comment]
Ok, we'll merge it and work on top. I believe there are many improvements that could be done (i.e it would be nice to have a better system based on interfaces to split the tests, and not a giant if/else on the operation mode).

Thanks for your work and enjoy your holidays!

**Review by p4u — APPROVED** [review]

**Pull request merged**